### ISSUE TITLE
GHA backend: support --submodules

### ISSUE THREAD

**Issue by RyanGlScott** [haskell-CI/haskell-ci#503]
As nice as `source-repository-package`s are, there are some situations where supporting `git` submodules would also be nice. `haskell-ci` has a `--submodules` flag for this purpose, but at the moment the GitHub Actions backend does not support it. At first, I thought supporting this flag would be as simple as adding `submodules: true` to the `checkout@v2` action, but when I tried this, I got [this error](https://github.com/RyanGlScott/gha-sandbox/runs/2289847432?check_suite_focus=true#step:10:32):

```
The repository will be downloaded using the GitHub REST API
To create a local Git repository instead, add Git 2.18 or higher to the PATH
Error: Input 'submodules' not supported when falling back to download using the GitHub REST API. To create a local Git repository instead, add Git 2.18 or higher to the PATH.
```

Darn. It looks like we might have to resort to a manual `git submodule update --init` to do this.

**Comment by RyanGlScott** [comment]
Alas, `git submodule update --init` doesn't appear to work either due to extremely weird behavior in an older version of `git` that `buildpack-deps:bionic` uses. See actions/checkout#335. I was only able to make things work by installing a newer version of `git` using an Ubuntu PPA, per the advice in [this comment](https://github.com/actions/checkout/issues/335#issuecomment-763159788). If that sounds reasonable, I can open a PR.

**Comment by phadej** [comment]
I'll be principled and say that `actions/checkout` should first fix their README that they require some minimal `git` version for some functionality. I'm not adding complexity which cannot be traced to official documentation. Issues aren't, as their preconditions may change in the future, e.g. `checkout` requiring newer `git` later on.

TL;DR, the patch should come with a link to official docs why it's made they way it is.

**Comment by RyanGlScott** [comment]
The [README for `actions/checkout`](https://github.com/actions/checkout/blob/25a956c84d5dd820d28caab9f86b8d183aeeff3d/README.md) says this:

> When Git 2.18 or higher is not in your PATH, falls back to the REST API to download the files.

So it's possible that the limitation really lies in GitHub's REST API, not `actions/checkout` itself.

**Comment by RyanGlScott** [comment]
In fact, that's almost assuredly what's happening. The [code in `actions/checkout` ](https://github.com/actions/checkout/blob/25a956c84d5dd820d28caab9f86b8d183aeeff3d/src/github-api-helper.ts#L116-L138) is using [this REST endpoint](https://docs.github.com/en/rest/reference/repos#download-a-repository-archive-tar) to download a `git` repo as a tarball which is later extracted. Based off my limited experimentation with this REST API, these tarballs do not come with `.git` folders, so `git` commands (e.g., `git submodule update --init`) simply do not work there.

In summary, I don't see any way to make this work without upgrading the `git` version.

**Comment by phadej** [comment]
Sorry, I'm not happy with additional complexity. Use `ubuntu: focal`. `bionic`'s end-of-support is in two years, so maybe a good idea to give it a try.

**Comment by RyanGlScott** [comment]
OK, I'd be willing to use a newer Ubuntu version. However, this poses a design question: what happens if a user tries to combine `--submodules` with `--distribution=xenial` or `--distribution=bionic`? Should `haskell-ci` throw an error? Or should it just accept whatever the user tells it?

Relatedly, https://launchpad.net/~hvr/+archive/ubuntu/ghcjs only supports Ubuntu 16.04 (Xenial) and 18.04 (Bionic), so it won't work with 20.04 (Focal). What should happen if a user tries enabling GHCJS on Focal?

**Comment by RyanGlScott** [comment]
I implemented the "throw an error" design in #504.
